crates/backend/src/codegen.rs

```diff
@@ -1135,6 +1138,135 @@ impl<'a> TryToTokens for ast::Module {
     }
 }
 
+impl ToTokens for ast::Dictionary {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let vis = &self.vis;
+        let name = &self.name;
+        let mut methods = TokenStream::new();
+        for field in self.fields.iter() {
+            field.to_tokens(&mut methods);
+        }
+        let required_names = &self.fields.iter()
+            .filter(|f| f.required)
+            .map(|f| &f.name)
+            .collect::<Vec<_>>();
+        let required_types = &self.fields.iter()
+            .filter(|f| f.required)
+            .map(|f| &f.ty)
+            .collect::<Vec<_>>();
+        let required_names2 = required_names;
+        let required_names3 = required_names;
+
+        let const_name = Ident::new(&format!("_CONST_{}", name), Span::call_site());
+        (quote! {
+            #[derive(Clone, Debug)]
+            #vis struct #name {
+                obj: ::js_sys::Object,
+            }
+
+            impl #name {
+                #vis fn new(#(#required_names: #required_types),*) -> #name {
+                    let mut _ret = #name { obj: ::js_sys::Object::new() };
+                    #(_ret.#required_names2(#required_names3);)*
+                    return _ret
+                }
+
+                #methods
+            }
+
+            #[allow(bad_style)]
+            const #const_name: () = {
+                use js_sys::Object;
+                use wasm_bindgen::describe::WasmDescribe;
+                use wasm_bindgen::convert::*;
+                use wasm_bindgen::JsValue;
+
+                // interop w/ JsValue
+                impl From<#name> for JsValue {
+                    fn from(val: #name) -> JsValue {
+                        val.obj.into()
+                    }
+                }
+                impl AsRef<JsValue> for #name {
+                    fn as_ref(&self) -> &JsValue { self.obj.as_ref() }
+                }
+                impl AsMut<JsValue> for #name {
+                    fn as_mut(&mut self) -> &mut JsValue { self.obj.as_mut() }
+                }
+
+                // interop w/ Object
+                impl From<#name> for Object {
```

Comment:
I don't think there is any guarantee that we could do this with the host bindings proposal. In particular, I don't see any mention that WebIDL dictionaries have to be ECMAScript Objects. (This is really more of a concern for the AsRef and AsMut impls than the conversion impl, but I think we should implement all or none.)